### ISSUE TITLE
[FEA] Refactor `concatenate` APIs

### ISSUE THREAD

**Issue by ttnghia** [rapidsai/cudf#7883]
**Is your feature request related to a problem? Please describe.**
Currently, there are a bunch of overloaded `concatenate` APIs that do different things, very different things. All of them have the same name `concatenate`. Some of them are located in their own namespace such as `lists::concatenate` or `string::concatenate`, which allow guessing their functionalities from their namespace. Unfortunately, just namespace is not enough. 

Overall, `concatenate` APIs can be classified into 2 types (correct me if I'm wrong): column-wise concatenation and row-wise concatenation. Some namespaces such as `strings` already have all those concatenate types. It is very confusing to just have `concatenate` APIs to use for both column-wise and row-wise concatenation.

**Describe the solution you'd like**
I would suggest to use prefix/suffix to differentiate different APIs' functionalities. For example, `hconcat` for horizontal (row-wise) concatenation and `vconcate` for vertical (column-wise) concatenation. Other better names other than `hconcate` and `vconcate` would be great.

**Comment by jrhemstad** [comment]
Can you provide examples? 

> concatenate_ws, or concatenate_rw

Definitely don't like those names. We should use human readable words in naming.

**Comment by ttnghia** [comment]
A very typical example is `strings::concatenate`:

In `cudf/strings/detail/concatenate.hpp`:
```
/**
 * @brief Returns a single column by vertically concatenating the given vector of
 * strings columns.
 *
 * ```
 * s1 = ['aa', 'bb', 'cc']
 * s2 = ['dd', 'ee']
 * r = concatenate_vertically([s1,s2])
 * r is now ['aa', 'bb', 'cc', 'dd', 'ee']
 * ```
 *
 * @param columns List of string columns to concatenate.
 * @param mr Device memory resource used to allocate the returned column's device memory.
 * @param stream CUDA stream used for device memory operations and kernel launches.
 * @return New column with concatenated results.
 */
std::unique_ptr<column> concatenate(
  host_span<column_view const> columns,
  rmm::cuda_stream_view stream        = rmm::cuda_stream_default,
  rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource());
```

In `cudf/strings/combine.hpp`:
```
/**
 * @brief Row-wise concatenates the given list of strings columns and
 * returns a single strings column result.
....
*/
std::unique_ptr<column> concatenate(
  lists_column_view const& lists_strings_columns,
  string_scalar const& separator      = string_scalar(""),
  string_scalar const& narep          = string_scalar("", false),
  rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource());
```

**Comment by ttnghia** [comment]
@davidwendt do you have any idea for this?

**Comment by davidwendt** [comment]
I always felt that adding a vector/column of rows to the end of another vector/column should be called `append`.
Initially I called this `concatenate_vertically` as you can see by the typo in the doxygen example.
Maybe ask our resident language person @codereport

**Comment by codereport** [comment]
Summary:
|Lang|a|b|
|:-:|:-:|:-:|
|APL|[catenate](http://microapl.com/apl_help/ch_020_020_490.htm)|[1st axis catenate](http://microapl.com/apl_help/ch_020_020_500.htm)|
|Julia|[hcat](https://docs.julialang.org/en/v1/base/arrays/#Base.hcat)|[vcat](https://docs.julialang.org/en/v1/base/arrays/#Base.vcat)|
|R|[cbind](https://www.rdocumentation.org/packages/base/versions/3.6.2/topics/cbind)|[rbind](https://www.rdocumentation.org/packages/SparkR/versions/2.4.6/topics/rbind)|

**Details**

We can't really borrow from APL - as they called them `catenate` and `1st axis catenate`.
![image](https://user-images.githubusercontent.com/36027403/113892559-1e0c2b00-9794-11eb-8deb-f53de0dbaf3e.png)
Here is an example:
![image](https://user-images.githubusercontent.com/36027403/113892801-6297c680-9794-11eb-9714-c52623a060f8.png)

```julia
julia> x = [[1;2;3],[4;5;6;]]
2-element Array{Array{Int64,1},1}:
 x[1, 2, 3]
 [4, 5, 6]

julia> x
2-element Array{Array{Int64,1},1}:
 [1, 2, 3]
 [4, 5, 6]

julia> x,x
([[1, 2, 3], [4, 5, 6]], [[1, 2, 3], [4, 5, 6]])

julia> hcat(x,x)
2×2 Array{Array{Int64,1},2}:
 [1, 2, 3]  [1, 2, 3]
 [4, 5, 6]  [4, 5, 6]

julia> vcat(x,x)
4-element Array{Array{Int64,1},1}:
 [1, 2, 3]
 [4, 5, 6]
 [1, 2, 3]
 [4, 5, 6]
```

**Comment by ttnghia** [comment]
From your snippets, I think `hconcat` and `vconcat` would be some good names to use.

**Comment by davidwendt** [comment]
I believe Pandas concat you specify the axis on which to perform the concat.
@kkraus14
I know @harrism likes to weigh on names. I think he talked my out of using `concatenate_vertically` and `concatenate_horizonatally` originally.

**Comment by kkraus14** [comment]
It does, but Pandas also allows you to run joins via the `concat` API, so...

**Comment by harrism** [comment]
I think the argument "namespace is not enough" is not relevant because `lists::concatenate` and `strings::concatenate` are detail APIs that are called by the public APIs, not the other way around.

**Comment by harrism** [comment]
`concatenate` today means `concatenate columns`. If you want to add a function to concatenate rows, call it what it does: `concatenate_rows`.  If you are concatenating multiple elements of each list into a string, that's something different, and should have a different name, probably. I think Python calls that "join", unfortunately.

**Comment by ttnghia** [comment]
> I think the argument "namespace is not enough" is not relevant because `lists::concatenate` and `strings::concatenate` are detail APIs that are called by the public APIs, not the other way around.

Then when you call `lists::concatenate`, is it for column-wise or row-wise concatenation? Having `lists::concatenate` and `lists::concatenate_rows` would be better, instead of just 2 overloaded `lists::concatenate` APIs.

**Comment by ttnghia** [comment]
@davidwendt Based on Mark's suggestion, should we change the concatenate APIs in `strings/combine.(hpp|cu)` into `strings::concatenate_rows`?

**Comment by davidwendt** [comment]
I'd rather not. What is the priority for this change?

**Comment by ttnghia** [comment]
Priority is low---it can be done by anybody at any time, but we should/must do it eventually to clean up the code and avoid further confusion.

**Comment by ttnghia** [comment]
@Everyone: I raised the issue with examples of strings APIs, but this issue is general. We are going to have similar issue for `lists/structs` as well, thus we must be clear about the name convention before adding more stuff.

**Comment by harrism** [comment]
I get the impression that not everyone is agreeing with you @ttnghia , so I don't think we have a concensus to make this a convention.  I read "I'd rather not" to mean "we should not make this change". Is that what you meant, @davidwendt ?

**Comment by davidwendt** [comment]
I think we can close this after #7929 is merged along with the changes merged in #7937
